### PR TITLE
Mvertens/caserun

### PR DESCRIPTION
**Pull request by mvertens**
pythonized machines/[template.case.run, template.st_archive and template.lt_archive] for both cesm and acme - ran regressions tests for cesm successfully - did not yet do so for acme

part of #149

### CODE REVIEW

File: cime_config/acme/machines/template.case.run [hunk 1]
```diff
@@ -1,416 +1,327 @@
...
+
+    # load the module environment...
+    env_module = EnvModule(mach, compiler, cimeroot, caseroot, mpilib, debug)
+    env_module.load_env_for_case()
+
+    # set environment variables
+    # currently need to set mpilib as an environment variable- otherwise
+    # the model dies at runtime 
+    # FIXME: is there a way to remove this requirement?
+    if mpilib == "mpi-serial":
```
Comment (jedwards4b) [inline cime_config/acme/machines/template.case.run]: I think that this is a yellowstone "feature", no other machine requires it.

File: cime_config/acme/machines/template.case.run [hunk 2]
```diff
@@ -1,416 +1,327 @@
...
+        del os.environ["MP_MPILIB"]
+    else:
+        os.environ["MPILIB"] = mpilib
+    if batchsubmit is None or len(batchsubmit) == 0:
```
Comment (jedwards4b) [inline cime_config/acme/machines/template.case.run]: Any idea what this is for?

File: cime_config/acme/machines/template.case.run [hunk 3]
```diff
@@ -1,416 +1,327 @@
...
-    chdir $rundir;
-    disposeLog($rundir,$caseroot, 'da' ,$lid,$logdir);
-    chdir $caseroot;
-}
+    # create the timing directories, optionally cleaning them if needed.
+    try:
+        os.stat(rundir)
+    except:
+        os.mkdir(rundir)       
+    if os.path.isdir(rundir + '/timing'):
```
Comment (jedwards4b) [inline cime_config/acme/machines/template.case.run]: use os.path.join

File: cime_config/acme/machines/template.case.run [hunk 4]
```diff
@@ -1,416 +1,327 @@
...
+###############################################################################
+def runModel(case):
+###############################################################################
 
+    # Set OMP_NUM_THREADS
+    num_threads = {{ thread_count }}
+    os.environ["OMP_NUM_THREADS"] = str(num_threads)
 
+    # Run the model
+    cmd = ("sleep 10")
```
Comment (jedwards4b) [inline cime_config/acme/machines/template.case.run]: we should try getting rid of the sleep slop

File: cime_config/acme/machines/template.case.run [hunk 5]
```diff
@@ -1,416 +1,327 @@
...
+        save_timing  = case.get_value("SAVE_TIMING")
+        rundir       = case.get_value("RUNDIR")
+        caseroot     = case.get_value("CASEROOT")
+        model        = case.get_value("MODEL")
+        
+        timingDir = os.path.join(caseroot, "timing")
+        if not os.path.isdir(timingDir):
+            os.makedirs(timingDir)
+
+        logger.info("Running timing script %s " %(os.path.join(caseroot, "Tools/getTiming")))
```
Comment (jedwards4b) [inline cime_config/acme/machines/template.case.run]: here you are sorta using os.path.join but not really ( you should never explicitly need the '/') - try:
os.path.join(caseroot, "Tools", "getTiming")

File: cime_config/acme/machines/template.case.run [hunk 6]
```diff
@@ -1,416 +1,327 @@
...
+        cmd = "%s -lid %s " %(os.path.join(caseroot,"Tools/getTiming"), lid)
+        run_cmd(cmd) 
+
+        # ave the timing files if desired
+        if save_timing:
+            shutil.move(os.path.join(rundir,"timing"), 
+                        os.path.join(rundir,"timing."+lid))
+
+        # compress relevant timing files
+        logger.info( "gzipping timing stats.." )
```
Comment (jedwards4b) [inline cime_config/acme/machines/template.case.run]: There is a native gzip in python - no need for a system call here

File: cime_config/acme/machines/template.case.run [hunk 7]
```diff
@@ -1,416 +1,327 @@
...
+        component = Component(config_file)
+        comps = [x.lower() for x in component.get_valid_model_components()]
+        comps = [x.replace('drv', 'cpl') for x in comps]
+        model = [case.get_value("MODEL")]
+        comps = comps + model
+
+        for comp in comps:
+            logfile = os.path.join(comp + '.log.' + lid)
+            if os.path.isfile(os.path.join(rundir,logfile)):
+                cmd = "gzip " + logfile
```
Comment (jedwards4b) [inline cime_config/acme/machines/template.case.run]: use python gzip

File: cime_config/acme/machines/template.case.run [hunk 8]
```diff
@@ -1,416 +1,327 @@
...
+        logger.debug( "running resubmit check %s " %cmd)
+        rc, output, error = run_cmd(cmd, ok_to_fail=True, from_dir=caseroot)
+        if rc != 0:
+            logger.warn("could not run %s submit script, rc=%s, output=%s, error=%s" %(model,rc,output,error))
+
+###############################################################################
+def DoDataAssimilation(case, da_script, lid):
+###############################################################################
+
+    cmd = "da_script 1> da.log.%s 2>&1" %(lid)
```
Comment (jedwards4b) [inline cime_config/acme/machines/template.case.run]: this won't work - da_script is a variable but you are putting it in cmd as a string

File: cime_config/acme/machines/template.lt_archive [hunk 9]
```diff
@@ -1,177 +1,117 @@
...
+    os.chdir(workdir)
+
+_CIMEROOT = os.environ.get("CIMEROOT")
+if _CIMEROOT is None:
+    raise exc_type("ERROR: must set CIMEROOT environment variable")
+
+_LIBDIR = os.path.join(_CIMEROOT, "scripts", "Tools")
+sys.path.append(_LIBDIR)
+
+from standard_script_setup          import *
```
Comment (jedwards4b) [inline cime_config/acme/machines/template.lt_archive]: Clean this up - I don't think you need most of these imports.

File: cime_config/acme/machines/template.lt_archive [hunk 10]
```diff
@@ -1,177 +1,117 @@
...
-        }
-        qx($resubmitcommand);
-        if($?) { $logger->warn("could not run CESM submit script, $! $?") ; }
-        chdir $cwd;
-    }
-}
-main(@ARGV) unless caller;
+    if runComplete and staComplete:
+        os.environ["DOUT_S_ROOT"]       = case.get_value("DOUT_S_ROOT")
+        os.environ["DOUT_L_MSROOT"]     = case.get_value("DOUT_L_MSROOT")
```
Comment (jedwards4b) [inline cime_config/acme/machines/template.lt_archive]: looks like an indentation problem here.

File: cime_config/acme/machines/template.lt_archive [hunk 11]
```diff
@@ -1,177 +1,117 @@
...
+               "lt_archive: run or st_archive is not yet complete or was not successful."
+               "Unable to perform long term archive...")
+
+    # document completion
+    if os.path.isfile(testlog):
+        with open(testlog, "a") as fd:
+            fd.write("lg_archive completed %s" %(time.strftime("%Y-%m-%d %H:%M:%S")))
+
+    # resubmit case if appropriate
+    resubmit = int(case.get_value("RESUBMIT"))
```
Comment (jedwards4b) [inline cime_config/acme/machines/template.lt_archive]: lt_archive should not need resubmit.

File: cime_config/acme/machines/template.st_archive [hunk 12]
```diff
@@ -1,78 +1,87 @@
...
+    os.chdir(workdir)
+
+_CIMEROOT = os.environ.get("CIMEROOT")
+if _CIMEROOT is None:
+    raise exc_type("ERROR: must set CIMEROOT environment variable")
+
+_LIBDIR = os.path.join(_CIMEROOT, "scripts", "Tools")
+sys.path.append(_LIBDIR)
+
+from standard_script_setup          import *
```
Comment (jedwards4b) [inline cime_config/acme/machines/template.st_archive]: clean up import list

File: cime_config/cesm/machines/template.case.run [hunk 13]
```diff
@@ -1,418 +1,327 @@
...
+    os.chdir(workdir)
+
+_CIMEROOT = os.environ.get("CIMEROOT")
+if _CIMEROOT is None:
+    raise exc_type("ERROR: must set CIMEROOT environment variable")
+
+_LIBDIR = os.path.join(_CIMEROOT, "scripts", "Tools")
+sys.path.append(_LIBDIR)
+
+from standard_script_setup          import *
```
Comment (jedwards4b) [inline cime_config/cesm/machines/template.case.run]: clean up import list

File: scripts/Tools/st_archive [hunk 14]
```diff
@@ -201,7 +201,6 @@ EOF
     exit(1);
 }
 
-
```
Comment (jedwards4b) [inline scripts/Tools/st_archive]: why is this file changed?

File: utils/perl5lib/Batch/BatchMaker.pm [hunk 15]
```diff
@@ -812,7 +812,7 @@ sub setCESMRun()
         }
     }
 
-    $self->{'mpirun'} .= "qx( " . $executableString . " " . $mpiargstring .  " " . $defaultrunsuffix . ");";
+    $self->{'mpirun'} .= "run_cmd( " . $executableString . " " . $mpiargstring .  " " . $defaultrunsuffix . ")";
```
Comment (jedwards4b) [inline utils/perl5lib/Batch/BatchMaker.pm]: I think that the run_cmd should be in template.case.run and not here.   In fact that is the way you are using it at line 142 of template.case.run

File: utils/python/CIME/batch_maker.py [hunk 16]
```diff
@@ -205,7 +205,7 @@ def _set_model_run(self):
 
         mpi_arg_string = " ".join(args.values())
 
-        self.mpirun = "%s %s %s" % (executable if executable is not None else "", mpi_arg_string, default_run_suffix)
+        self.mpirun = "cmd =\"%s %s %s \" " % (executable if executable is not None else "", mpi_arg_string, default_run_suffix)
```
Comment (jedwards4b) [inline utils/python/CIME/batch_maker.py]: what is this?   For debug?

File: utils/python/CIME/case_setup.py [hunk 17]
```diff
@@ -139,7 +139,7 @@ def case_setup(caseroot, clean=False, test_mode=False, reset=False):
         logger.info("Some files have been saved to %s" % backup_dir)
 
         with open("CaseStatus", "a") as fd:
-            fd.write("case.setup --clean %s\n" % time.strftime("%Y-%m-%d %H:%M:%S"))
+            fd.write("%s case.setup --clean \n" % time.strftime("%Y-%m-%d %H:%M:%S"))
```
Comment (jedwards4b) [inline utils/python/CIME/case_setup.py]: I think that we should put the CaseStatus append function in CIME.utils and use it uniformly throughout the code.

**Comment by jedwards4b** [comment]
I've reviewed this code, comments on files in acme/machines also apply to those in cesm/machines.

File: cime_config/acme/machines/template.case.run (hunk 1 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: So what do you suggest we do? I'm wary to remove it right now.

Comment (jedwards4b) [inline cime_config/acme/machines/template.case.run]: needed until we move on from yellowstone.  We should just update the note to explain better why it's there.

File: cime_config/acme/machines/template.case.run (hunk 2 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: My guess was that in the perl this was being used to run a mpi-serial job interactively. Thoughts?

File: cime_config/acme/machines/template.case.run (hunk 3 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: good point.

File: cime_config/acme/machines/template.case.run (hunk 4 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: Agreed. Do you want to do this as part of this pull request? Or should we just create an issue that this should be done in a cleanup phase.

File: cime_config/acme/machines/template.case.run (hunk 5 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: Good point. Old habits are hard to break.

File: cime_config/acme/machines/template.case.run (hunk 6 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: Thanks for pointing this out.

File: cime_config/acme/machines/template.case.run (hunk 8 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: We should definitely have some kind of data assimilation test - since the regular testing did not pick this up.

File: cime_config/acme/machines/template.case.run (hunk 4 above)
Comment (jedwards4b) [inline cime_config/acme/machines/template.case.run]: an issue would be fine

File: cime_config/acme/machines/template.case.run [hunk 18]
```diff
@@ -1,416 +1,327 @@
...
+    env_module.load_env_for_case()
+
+    # set environment variables
+    # currently need to set mpilib as an environment variable- otherwise
+    # the model dies at runtime 
+    # FIXME: is there a way to remove this requirement?
+    if mpilib == "mpi-serial":
+        del os.environ["MP_MPILIB"]
+    else:
+        os.environ["MPILIB"] = mpilib
```
Comment (jgfouca) [inline cime_config/acme/machines/template.case.run]: Please put a blank line to visually separate the previous if/else block from the next one.

File: cime_config/acme/machines/template.case.run [hunk 19]
```diff
@@ -1,416 +1,327 @@
...
-    system(" $da_script 1> da.log.$lid 2>&1");
-    sleep(1);
-    $logger->logdie ("$da_script failed") unless ($? == 0);
-    chdir $rundir;
-    disposeLog($rundir,$caseroot, 'da' ,$lid,$logdir);
-    chdir $caseroot;
-}
+    # create the timing directories, optionally cleaning them if needed.
+    try:
+        os.stat(rundir)
```
Comment (jgfouca) [inline cime_config/acme/machines/template.case.run]: Don't use exceptions to handle normal code flow. I think you want this:

    if not os.path.isdir(rundir):
        os.mkdir(rundir)

File: cime_config/acme/machines/template.case.run (hunk 4 above)
Comment (jgfouca) [inline cime_config/acme/machines/template.case.run]: I agree, sleep 10 is quite a penalty for the user. Why was it ever needed?

File: cime_config/acme/machines/template.case.run [hunk 20]
```diff
@@ -1,416 +1,327 @@
...
+        # ave the timing files if desired
+        if save_timing:
+            shutil.move(os.path.join(rundir,"timing"), 
+                        os.path.join(rundir,"timing."+lid))
+
+        # compress relevant timing files
+        logger.info( "gzipping timing stats.." )
+        timingfile = model + "_timing_stats." + lid
+        cmd = "gzip " + (os.path.join(caseroot, "timing", timingfile))
+        print "DEBUG: command is ",cmd
```
Comment (jgfouca) [inline cime_config/acme/machines/template.case.run]: Please remove this DEBUG print. Use logger.debug if you want.

File: cime_config/acme/machines/template.case.run [hunk 21]
```diff
@@ -1,416 +1,327 @@
...
+            logger.warn("could not run %s submit script, rc=%s, output=%s, error=%s" %(model,rc,output,error))
+
+###############################################################################
+def DoDataAssimilation(case, da_script, lid):
+###############################################################################
+
+    cmd = "da_script 1> da.log.%s 2>&1" %(lid)
+    run_cmd(cmd)
+    cmd = "sleep(1)"
+    run_cmd(cmd)
```
Comment (jgfouca) [inline cime_config/acme/machines/template.case.run]: Why sleep?

File: cime_config/acme/machines/template.case.run [hunk 22]
```diff
@@ -1,416 +1,327 @@
...
+
+    data_assimilation        = case.get_value("DATA_ASSIMILATION")
+    data_assimilation_cycles = case.get_value("DATA_ASSIMILATION_CYCLES")
+    data_assimilation_script = case.get_value("DATA_ASSIMILATION_SCRIPT")
+
+    # set up the LID 
+    lid = time.strftime("%y%m%d-%H%M%S")
+    os.environ["LID"] = lid
+
+    for cycle in range(0,data_assimilation_cycles):
```
Comment (jgfouca) [inline cime_config/acme/machines/template.case.run]: The zero as first argument to range is not necessary. Python assumes zero is the start if you only provide one argument.

File: cime_config/acme/machines/template.case.run [hunk 23]
```diff
@@ -1,416 +1,327 @@
...
+        files = Files()
+        config_file = files.get_value("CONFIG_DRV_FILE")
+        component = Component(config_file)
+        comps = [x.lower() for x in component.get_valid_model_components()]
+        comps = [x.replace('drv', 'cpl') for x in comps]
+        model = [case.get_value("MODEL")]
+        comps = comps + model
+
+        for comp in comps:
+            logfile = os.path.join(comp + '.log.' + lid)
```
Comment (jgfouca) [inline cime_config/acme/machines/template.case.run]: This os.path.join command isn't doing anything. You only provided one argument.

File: cime_config/acme/machines/template.lt_archive (hunk 9 above)
Comment (jgfouca) [inline cime_config/acme/machines/template.lt_archive]: pylint can show unused imports

File: cime_config/acme/machines/template.lt_archive [hunk 24]
```diff
@@ -1,177 +1,117 @@
...
+    statusfile = os.path.join(caseroot,"CaseStatus") 
+    if os.path.isfile(statusFile):
+        if 'run SUCCESSFUL' in open(statusFile).read():
+            runComplete = True
+
+    statusComplete = False
+    statusfile = os.path.join(caseroot,"stArchiveStatus")
+    if os.path.isfile(statusFile):
+        if 'st_archive process complete' in open(statusFile).read():
+            runComplete = True
```
Comment (jgfouca) [inline cime_config/acme/machines/template.lt_archive]: I think this should be statusComplete = True

File: cime_config/acme/machines/template.lt_archive [hunk 25]
```diff
@@ -1,177 +1,117 @@
...
+    # max number of threads needed by scripts
+    os.environ["maxthrds"] = 1
+
+    # document start
+    if os.path.isfile(testlog):
+        with open(testlog, "a") as fd:
+            fd.write("lt_archive starting %s" %(time.strftime("%Y-%m-%d %H:%M:%S")))
+    logger.info("lt_archive starting %s" %(time.strftime("%Y-%m-%d %H:%M:%S")))
+
+    # determine status of run and short term archiving
```
Comment (jgfouca) [inline cime_config/acme/machines/template.lt_archive]: The following two code blocks could be a good candidate for a simple utility function. Something like:

     def does_file_have_string(filepath, text):
         return os.path.isfile(filepath) and text in open(filepath).read()

File: cime_config/acme/machines/template.lt_archive [hunk 26]
```diff
@@ -1,177 +1,117 @@
...
-        chdir $cwd;
-    }
-}
-main(@ARGV) unless caller;
+    if runComplete and staComplete:
+        os.environ["DOUT_S_ROOT"]       = case.get_value("DOUT_S_ROOT")
+        os.environ["DOUT_L_MSROOT"]     = case.get_value("DOUT_L_MSROOT")
+	os.environ["DOUT_L_HPSS_ACCNT"] = case.get_value("DOUT_L_HPSS_ACCNT")
+
+	lid = time.strftime("%y%m%d-%H%M%S")
```
Comment (jgfouca) [inline cime_config/acme/machines/template.lt_archive]: Uneven indent. pylint can find these

File: cime_config/acme/machines/template.lt_archive [hunk 27]
```diff
@@ -1,177 +1,117 @@
...
+
+    # document completion
+    if os.path.isfile(testlog):
+        with open(testlog, "a") as fd:
+            fd.write("lg_archive completed %s" %(time.strftime("%Y-%m-%d %H:%M:%S")))
+
+    # resubmit case if appropriate
+    resubmit = int(case.get_value("RESUBMIT"))
+    if resubmit > 0:                   
+	cmd = os.path.join(caseroot, "case.submit") + " -resubmit -scriptname case.st_archive"
```
Comment (jgfouca) [inline cime_config/acme/machines/template.lt_archive]: uneven indent

File: cime_config/cesm/machines/template.case.run [hunk 28]
```diff
@@ -1,418 +1,327 @@
-#! /usr/bin/env perl
-#------------------------------------------------------------------------------
-# Batch system directives
-#------------------------------------------------------------------------------
+#!/usr/bin/env python2
+
+"""
+template to create a case run script
+"""
+
```
Comment (jgfouca) [inline cime_config/cesm/machines/template.case.run]: I assume this is just a copy of what I reviewed above.

File: utils/python/CIME/case_setup.py (hunk 17 above)
Comment (jgfouca) [inline utils/python/CIME/case_setup.py]: Yes please.

**Comment by jgfouca** [comment]
Review complete. Looks real good overall.

File: cime_config/acme/machines/template.case.run (hunk 1 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: Done - removed all sleep references.

File: cime_config/acme/machines/template.case.run (hunk 7 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: Done.

File: cime_config/acme/machines/template.case.run (hunk 8 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: Done.

File: cime_config/acme/machines/template.lt_archive (hunk 9 above)
Comment (mvertens) [inline cime_config/acme/machines/template.lt_archive]: Done. Unfortunately, pylint is not installed in any of our systems. I'll install it on my laptop - but for now I did this manually.

File: cime_config/acme/machines/template.lt_archive (hunk 10 above)
Comment (mvertens) [inline cime_config/acme/machines/template.lt_archive]: There was no indentation problem.

File: cime_config/acme/machines/template.st_archive (hunk 12 above)
Comment (mvertens) [inline cime_config/acme/machines/template.st_archive]: Done

File: cime_config/cesm/machines/template.case.run (hunk 13 above)
Comment (mvertens) [inline cime_config/cesm/machines/template.case.run]: Done - but most of these imports are being used here.

File: scripts/Tools/st_archive (hunk 14 above)
Comment (mvertens) [inline scripts/Tools/st_archive]: For my readability in tackling its pythonization next.

File: utils/perl5lib/Batch/BatchMaker.pm (hunk 15 above)
Comment (mvertens) [inline utils/perl5lib/Batch/BatchMaker.pm]: Done. I am not using BatchMaker.pm. Jim will be getting rid of as much of this as possible in his PR.

File: utils/python/CIME/case_setup.py (hunk 17 above)
Comment (mvertens) [inline utils/python/CIME/case_setup.py]: Done

File: cime_config/acme/machines/template.case.run (hunk 18 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: Done

File: cime_config/acme/machines/template.case.run (hunk 19 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: Done.

File: cime_config/acme/machines/template.case.run (hunk 21 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: Done.

File: cime_config/acme/machines/template.case.run (hunk 22 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: Done

File: cime_config/acme/machines/template.case.run (hunk 23 above)
Comment (mvertens) [inline cime_config/acme/machines/template.case.run]: Fixed.

File: cime_config/acme/machines/template.lt_archive (hunk 24 above)
Comment (mvertens) [inline cime_config/acme/machines/template.lt_archive]: Done. Right. We need to add a fast long term archiving test. Jim is thinking about this as well.

File: cime_config/acme/machines/template.lt_archive (hunk 25 above)
Comment (mvertens) [inline cime_config/acme/machines/template.lt_archive]: Done. Great idea. Thanks.

File: cime_config/acme/machines/template.lt_archive (hunk 26 above)
Comment (mvertens) [inline cime_config/acme/machines/template.lt_archive]: This is odd. My script does not show this. I've downloaded pylint and will use this in the future.

File: cime_config/acme/machines/template.lt_archive (hunk 27 above)
Comment (mvertens) [inline cime_config/acme/machines/template.lt_archive]: Used pylint - okay
